### PR TITLE
Add GDS support for vertex property scanning

### PR DESCRIPTION
**Pull request by benjaminwinger**
I think the new `NodeGroup::scan` function may still be buggy. I have attempted to merge the two implementations, but have encountered issues with the result.

I renamed `NodeGroupScanState::numScannedRows` to `nextRowToScan` since they are functionally equivalent and nothing was ever using `numScannedRows` as a size, just as an indication of the next row to scan. There was also a field called `nextRowToScan` already in `CSRNodeGroupScanState`, so that was just removed and the other field is now used in its place.

### CODE REVIEW

**Review by semihsalihoglu-uw — CHANGES_REQUESTED** [review]
I have a few comments but it looks like you want to work on this PR a bit more. Especially the following comment you made makes me think it's not yet ready to merge: "I think the new NodeGroup::scan function may still be buggy. I have attempted to merge the two implementations, but have encountered issues with the result." 

Let's try to get back to this PR asap and merge because @acquamarin depends on this.

File: src/storage/store/node_group.cpp [hunk 1]
```diff
@@ -193,11 +196,107 @@ NodeGroupScanResult NodeGroup::scan(Transaction* transaction, TableScanState& st
...
             numRowsToScan);
     }
-    const auto startRow = nodeGroupScanState.numScannedRows;
-    nodeGroupScanState.numScannedRows += numRowsToScan;
+    const auto startRow = nodeGroupScanState.nextRowToScan;
+    nodeGroupScanState.nextRowToScan += numRowsToScan;
     return NodeGroupScanResult{startRow, numRowsToScan};
 }
 
+NodeGroupScanResult NodeGroup::scan(Transaction* transaction, TableScanState& state,
```
[the remark this answers is not in the document]
Comment (benjaminwinger) [reply src/storage/store/node_group.cpp:204]: Yes, I can try and add to the new tests. I'd been trying to merge this new function with the existing one so that it would be better covered, but I had run into some problems with that.

File: src/graph/on_disk_graph.cpp [hunk 2]
```diff
@@ -305,5 +312,93 @@ bool OnDiskGraphScanStates::next() {
...
+
+std::unique_ptr<VertexScanState> OnDiskGraph::prepareVertexScan(common::table_id_t tableID,
+    const std::vector<std::string>& propertiesToScan) {
+    return std::make_unique<OnDiskGraphVertexScanState>(*context, tableID, propertiesToScan);
+}
+
+void OnDiskGraphVertexScanState::startScan(common::offset_t beginOffset,
+    common::offset_t endOffsetExclusive) {
+    numNodesScanned = 0;
+    // TODO(bmwinger): this might not be necessary here
```
[the remark this answers is not in the document]
Comment (benjaminwinger) [reply src/graph/on_disk_graph.cpp:384]: Yes, they were mostly just stuff that I was meaning to get back to in this PR. I'll see how many of them I can resolve.

File: src/storage/store/node_group.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (semihsalihoglu-uw) [reply src/storage/store/node_group.cpp:204]: Have you added tests for these? I still see many of these lines not tested.

[the remark this answers is not in the document]
Comment (benjaminwinger) [reply src/storage/store/node_group.cpp:208]: I didn't actually add any new tests. The bulk of what wasn't covered, was [this block](https://app.codecov.io/gh/kuzudb/kuzu/pull/4453#e5237d6f81b6af3f8b03ca119c6a2d70-R208), which was duplicated from elsewhere in the file. I've since merged them into a single function [here](https://github.com/kuzudb/kuzu/pull/4453/files#diff-776a83f6364f9ed72b1143204edde01675e100adc037f35e605dd18c038321a4R141-R169), but code-cov doesn't appear to actually update the coverage report when the PR is updated, so it still is showing those original lines as uncovered.

File: src/storage/store/node_group.cpp [hunk 3]
```diff
@@ -130,19 +133,48 @@ void NodeGroup::initializeScanState(Transaction*, const UniqLock& lock,
     TableScanState& state) const {
     auto& nodeGroupScanState = *state.nodeGroupScanState;
     nodeGroupScanState.chunkedGroupIdx = 0;
-    nodeGroupScanState.numScannedRows = 0;
+    nodeGroupScanState.nextRowToScan = 0;
     ChunkedNodeGroup* firstChunkedGroup = chunkedGroups.getFirstGroup(lock);
     initializeScanStateForChunkedGroup(state, firstChunkedGroup);
 }
 
+void applySemiMaskFilter(const TableScanState& state, row_idx_t numRowsToScan,
```
Comment (semihsalihoglu-uw) [inline src/storage/store/node_group.cpp:141]: How is this function related to GDS vertex property scanning?

**Review by semihsalihoglu-uw — APPROVED** [review]

File: src/storage/store/node_group.cpp (hunk 3 above)
Comment (benjaminwinger) [reply src/storage/store/node_group.cpp:141]: It's not, but the vertex property scanning needed a new overload of `NodeGroup::scan`, and I thought that new function should probably also support the semi masks and moved that code into a function so that it doesn't need to be duplicated.
